Improve theming documentation (#1615)
vaadin/docs: articles/styling/advanced/importing-stylesheets.adoc

---
title: Importing Standalone Style Sheets
description: Describes how to import standalone stylesheets using the @CssImport or @StyleSheet annotations.
order: 30
---

= Importing Standalone CSS Style Sheets

The easiest way to customize the look and feel of your application is to use the <<{articles}/styling/getting-started#, theming>> mechanism.
This mechanism automatically takes care of importing the CSS style sheets that you place under a specific folder in your project.
You should, therefore, consider using this mechanism as it is the easiest way to customize your application's styling.

However, you can also import standalone CSS style sheets manually using the `@CssImport` and the `@StyleSheet` annotations.

You should be familiar with <<{articles}/styling/custom-theme/style-scopes#,Style Scopes>> to know whether you need to import a style sheet to the global scope or to a component scope.
To learn where CSS files should be placed in your project, see <<../advanced/loading-resources#,Loading Resources>>.

[[importing.bundled]]
== Bundled Style Sheets

> Use the `@CssImport` annotation to import style sheets which are in your application source folder.

Reviewer: ```suggestion
Use the `@CssImport` annotation to import style sheets in the application's source code folder.

```

Bundled style sheets are included in the application bundle during a <<{articles}/production#,production build>>, together with other client-side resources.

Bundling is recommended for styles that change together with the application logic or component implementations, as the browser can cache them as a single unit of related resources.

[source,java]
----
// Import a style sheet into the global scope
@CssImport("./styles/shared-styles.css")
public class MyApplication extends Div {
}
----

> The `@CssImport` annotation can also be used to import component-specific style sheets.

Reviewer: I wonder if we should use a different term here than _"component-specific style sheets"_? I'm worried this can be a bit ambiguous and not sufficiently distinct from e.g. using component-specific selectors in global stylesheets etc. Not sure what would be better, maybe _"component shadow dom style sheets"_ or _"styles injected into components"? Just throwing the thought out there.

Author: How about?
_The `@CssImport` annotation can also be used to import CSS style sheets that target the shadow DOM of the components._

Reviewer: that's good yeah!

Author: Fixed.

[source,java]
----
// Import a style sheet into the local scope of the TextField component
@CssImport(value = "./styles/text-field.css",
           themeFor = "vaadin-text-field")
----

[[importing.external]]
== External Style Sheets

The `@StyleSheet` annotation can be used to import style sheets from an external URL, or from a URL within your application.
The latter type of URLs are prefixed with `context://`, which points to the root of the public resources folder of your application.

External/linked style sheets can be used to import styles without including the contents in the application bundle.
This allows the browser to load and cache the style sheet independently from the rest of the application.

> External style sheets need to be accessible from a URL, and should be placed in the public resource folder in your web application.

Reviewer: maybe an example of where you'd typically find that folder?

Author: 👍 added

They can also come from outside your web application, for example from a different domain or a content delivery network (CDN).

[source,java]
----
@StyleSheet("context://custom-font.css")
@StyleSheet("https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css")
public class MyApplication extends Div {
}
----

.External CSS can override your styles
[CAUTION]

> There is nothing to prevent the CSS rules in external style sheets from overriding component or view styles.

Reviewer: Is this really a concern worth pointing out?

Author: Maybe not. Removed.

[source,java]
----
// Import a style sheet into the global scope
@CssImport("./styles/shared-styles.css")
public class MyApplication extends Div {
}
----

The `@CssImport` annotation can also be used to import CSS style sheets that target the shadow DOM of the components.

[source,java]
----
// Import a style sheet into the local scope of the TextField component
@CssImport(value = "./styles/text-field.css",
           themeFor = "vaadin-text-field")
----

[[importing.external]]
== External Style Sheets

The `@StyleSheet` annotation can be used to import style sheets from an external URL, or from a URL within your application.
The latter type of URLs are prefixed with `context://`, which points to the root of the public resources folder of your application.

External/linked style sheets can be used to import styles without including the contents in the application bundle.
This allows the browser to load and cache the style sheet independently from the rest of the application.

External style sheets need to be accessible from a URL, and should be placed in the public resource folder in your web application.

> For example, in a Spring Boot application, if the style sheet file is called `shared-styles.css`, then it should be placed in the `src/main/resources/META-INF/resources` folder, and it can be imported using the `@StyleSheet` annotation as follows:

Reviewer 2: ```suggestion
For example, in a Spring Boot application, if the style sheet file is called [filename]#shared-styles.css#, then it should be placed in the `src/main/resources/META-INF/resources` folder, and it can be imported using the `@StyleSheet` annotation as follows:

```
